pkg/store/cache/caching_bucket.go

```diff
@@ -26,92 +29,178 @@ import (
 )
 
 const (
+	config = "config"
+
 	originCache  = "cache"
 	originBucket = "bucket"
-)
 
-type CachingBucketConfig struct {
-	// Basic unit used to cache chunks.
-	ChunkSubrangeSize int64 `yaml:"chunk_subrange_size"`
+	existsTrue  = "true"
+	existsFalse = "false"
 
-	// Maximum number of GetRange requests issued by this bucket for single GetRange call. Zero or negative value = unlimited.
-	MaxChunksGetRangeRequests int `yaml:"max_chunks_get_range_requests"`
-
-	// TTLs for various cache items.
-	ChunkObjectSizeTTL time.Duration `yaml:"chunk_object_size_ttl"`
-	ChunkSubrangeTTL   time.Duration `yaml:"chunk_subrange_ttl"`
-}
+	operationGet        = "get"
+	operationGetRange   = "getrange"
+	operationIter       = "iter"
+	operationExists     = "exists"
+	operationObjectSize = "objectsize"
+)
 
-func DefaultCachingBucketConfig() CachingBucketConfig {
-	return CachingBucketConfig{
-		ChunkSubrangeSize:         16000, // Equal to max chunk size.
-		ChunkObjectSizeTTL:        24 * time.Hour,
-		ChunkSubrangeTTL:          24 * time.Hour,
-		MaxChunksGetRangeRequests: 3,
-	}
-}
+var errObjNotFound = errors.Errorf("object not found")
 
 // Bucket implementation that provides some caching features, using knowledge about how Thanos accesses data.
 type CachingBucket struct {
 	objstore.Bucket
 
-	cache cache.Cache
+	logger log.Logger
 
-	config CachingBucketConfig
+	requestedGetRangeBytes *prometheus.CounterVec
+	fetchedGetRangeBytes   *prometheus.CounterVec
+	refetchedGetRangeBytes *prometheus.CounterVec
 
-	logger log.Logger
+	operationConfigs  map[string][]*operationConfig
+	operationRequests *prometheus.CounterVec
+	operationHits     *prometheus.CounterVec
+}
+
+// Generic config for single operation.
+type operationConfig struct {
+	configName string
+	matcher    func(name string) bool
+	cache      cache.Cache
+	opConfig   interface{}
+}
+
+// Operation-specific configs.
+type iterConfig time.Duration
+
+type existsConfig struct {
+	existsTTL, doesntExistTTL time.Duration
+}
 
-	requestedChunkBytes prometheus.Counter
-	fetchedChunkBytes   *prometheus.CounterVec
-	refetchedChunkBytes *prometheus.CounterVec
+type getConfig struct {
+	contentTTL, existsTTL, doesntExistTTL time.Duration
+}
 
-	objectSizeRequests prometheus.Counter
-	objectSizeHits     prometheus.Counter
+type getRangeConfig struct {
+	subrangeSize               int64
+	objectSizeTTL, subrangeTTL time.Duration
+	maxSubRequests             int
 }
 
-func NewCachingBucket(b objstore.Bucket, c cache.Cache, chunks CachingBucketConfig, logger log.Logger, reg prometheus.Registerer) (*CachingBucket, error) {
+type objectSizeCfg time.Duration
+
+// NewCachingBucket creates caching bucket with no configuration. Various "Cache*" methods configure
+// this bucket to cache results of individual bucket methods. Configuration must be set before
```

Comment:
I think it would be type safe if we could pass configuration ONLY via constructor. Any blocker for that? Especially if we cannot change it in runtime.

Reply:
Sounds good, I will make it work that way.